##Key Features of CrewAI:

####Role-Based Architecture: Agents are assigned distinct roles and goals, allowing for specialized task execution.
####Agent Orchestration: Facilitates the coordination of multiple agents, ensuring they work cohesively towards common objectives.
####Tool Integration: Supports the incorporation of various tools, extending agent capabilities to interact with external systems and perform diverse functions.
####Scalability: Designed to handle both simple and complex multi-agent systems, making it adaptable to various project sizes and complexities

# Task
Create a multi-agent system using the CrewAI library to simulate a grocery shop workflow, involving an 'Inventory Manager' to track stock and a 'Personal Shopper' to assist with customer selections, and execute a sample workflow to demonstrate their interaction.

## Install Dependencies

### Subtask:
Install the `crewai` library and any necessary dependencies required to run the multi-agent system.


**Reasoning**:
Install the `crewai` and `langchain_openai` libraries as required by the subtask instructions.



In [1]:
!pip install crewai langchain_openai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.3/89.3 kB 4.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 2.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 68.6/68.6 kB 3.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.5/40.5 kB 2.9 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of langchain-openai to determine which version is compatible with other requirements. This could take a while.
INFO: pip is still looking at multiple versions of langchain-openai to determine which version is compatible with other requirements. This could take a while.
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 3.2 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of i

**Reasoning**:
Verify the successful installation of the crewai library by importing the Agent, Task, and Crew classes as specified in the subtask instructions.



In [2]:
try:
    from crewai import Agent, Task, Crew
    print("Successfully imported Agent, Task, and Crew from crewai.")
except ImportError as e:
    print(f"Import failed: {e}")

Successfully imported Agent, Task, and Crew from crewai.


## Define Grocery Shop Agents

### Subtask:
Create the 'Inventory Manager' and 'Personal Shopper' agents with specific roles, goals, and backstories using the CrewAI framework.


**Reasoning**:
In this step, we will define the two agents for our grocery shop system: the 'Inventory Manager' and the 'Personal Shopper'. We'll begin by setting up the environment with the necessary API key (using a placeholder) and then initialize the agents with specific roles, goals, and backstories as outlined in the subtask instructions. Both agents will have `verbose=True` to allow us to see their internal reasoning process during execution.

**Reasoning**:
Instantiate the 'Inventory Manager' and 'Personal Shopper' agents with their respective roles, goals, and backstories as specified in the subtask instructions, after setting up the necessary environment variables.



In [3]:
import os
from crewai import Agent

# Set the OpenAI API key environment variable
os.environ["OPENAI_API_KEY"] = "your_api_key_here"

# Instantiate the Inventory Manager agent
inventory_manager = Agent(
    role='Inventory Manager',
    goal='Provide accurate product availability data and track stock levels.',
    backstory='A logistics expert with years of experience in retail management, ensuring precision and attention to detail in stock tracking.',
    verbose=True
)

# Instantiate the Personal Shopper agent
personal_shopper = Agent(
    role='Personal Shopper',
    goal='Curate shopping lists based on customer needs and assist with selections.',
    backstory='A friendly and knowledgeable grocery enthusiast with deep product knowledge, dedicated to helping customers find exactly what they need.',
    verbose=True
)

print("Agents 'Inventory Manager' and 'Personal Shopper' have been successfully instantiated.")

Agents 'Inventory Manager' and 'Personal Shopper' have been successfully instantiated.


## Define Grocery Shop Tasks

### Subtask:
Define specific tasks for the 'Inventory Manager' and 'Personal Shopper' agents using the CrewAI Task class.


**Reasoning**:
Define the `check_inventory_task` and `personal_shopper_task` using the `Task` class from the `crewai` library. These tasks will specify the descriptions, expected outputs, and assigned agents as outlined in the subtask instructions.

**Reasoning**:
In this step, we will define the specific tasks for our 'Inventory Manager' and 'Personal Shopper' agents using the `Task` class from the `crewai` library, as specified in the subtask instructions.



In [4]:
from crewai import Task

# Define the task for the Inventory Manager
check_inventory_task = Task(
    description="Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.",
    expected_output="A list indicating the availability and stock levels of the requested products.",
    agent=inventory_manager
)

# Define the task for the Personal Shopper
personal_shopper_task = Task(
    description="Create a customized shopping list for a customer looking for breakfast essentials, utilizing the stock information provided by the inventory manager.",
    expected_output="A curated shopping list with specific product recommendations for the customer.",
    agent=personal_shopper
)

print("Tasks 'check_inventory_task' and 'personal_shopper_task' have been successfully defined.")

Tasks 'check_inventory_task' and 'personal_shopper_task' have been successfully defined.


## Assemble and Execute Crew

### Subtask:
Initialize the Crew with the defined agents and tasks, and execute a sample workflow to simulate a grocery shopping interaction.


**Reasoning**:
In this final step, we will assemble the multi-agent system by instantiating the `Crew` class with our 'Inventory Manager' and 'Personal Shopper' agents and their respective tasks. We will configure the workflow to run sequentially using `Process.sequential`, ensuring that the inventory check is completed before the personal shopper creates the list. Finally, we will execute the workflow using the `kickoff()` method and display the results.

**Reasoning**:
In this step, we will instantiate the `Crew` class with the previously defined agents and tasks. We will set the process to `Process.sequential` to ensure the tasks are executed in order, and then call the `kickoff()` method to start the workflow and display the final result.



In [5]:
from crewai import Crew, Process

# Instantiate the Crew with agents and tasks
grocery_crew = Crew(
    agents=[inventory_manager, personal_shopper],
    tasks=[check_inventory_task, personal_shopper_task],
    process=Process.sequential,
    verbose=True
)

# Execute the workflow
result = grocery_crew.kickoff()

print("\n\n########################")
print("## Final Result:")
print("########################\n")
print(result)

╭─────────────────────────────────────────── 🚀 Crew Execution Started ───────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Started                                                                                         │
│  Name:                                                                                                          │
│  crew                                                                                                           │
│  ID:                                                                                                            │
│  526a472c-7322-4412-8afa-4dd5ef5592ef                                                                           │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── 📋 Task Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Started                                                                                                   │
│  Name: Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.  │
│  ID: 138161f3-a087-4211-b77a-6d684bc4a480                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Inventory Manager                                                                                       │
│                                                                                                                 │
│  Task: Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.  │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

ERROR:root:OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
ERROR:root:OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}


╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided:            │
│  your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':        │
│  'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}                                            │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[CrewAIEventsBus] Warning: Event pairing mismatch. 'llm_call_failed' closed 'agent_execution_started' (expected 
'llm_call_started')

An unknown error occurred. Please check the details below.
Error details: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
An unknown error occurred. Please check the details below.
Error details: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}


╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided:            │
│  your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':        │
│  'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}                                            │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Inventory Manager                                                                                       │
│                                                                                                                 │
│  Task: Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.  │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

ERROR:root:OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
ERROR:root:OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}


╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided:            │
│  your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':        │
│  'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}                                            │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[CrewAIEventsBus] Warning: Event pairing mismatch. 'llm_call_failed' closed 'agent_execution_started' (expected 
'llm_call_started')

An unknown error occurred. Please check the details below.
Error details: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
An unknown error occurred. Please check the details below.
Error details: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}


╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided:            │
│  your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':        │
│  'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}                                            │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Inventory Manager                                                                                       │
│                                                                                                                 │
│  Task: Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.  │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

ERROR:root:OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
ERROR:root:OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}


╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided:            │
│  your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':        │
│  'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}                                            │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[CrewAIEventsBus] Warning: Event pairing mismatch. 'llm_call_failed' closed 'agent_execution_started' (expected 
'llm_call_started')

An unknown error occurred. Please check the details below.
Error details: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
An unknown error occurred. Please check the details below.
Error details: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}


[CrewAIEventsBus] Warning: Event pairing mismatch. 'agent_execution_error' closed 'task_started' (expected 
'agent_execution_started')

╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided:            │
│  your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':        │
│  'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}                                            │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[CrewAIEventsBus] Warning: Event pairing mismatch. 'task_failed' closed 'crew_kickoff_started' (expected 
'task_started')

[CrewAIEventsBus] Warning: Ending event 'crew_kickoff_failed' emitted with empty scope stack. Missing starting 
event?

╭──────────────────────────────────────────────── 📋 Task Failure ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Failed                                                                                                    │
│  Name:                                                                                                          │
│  Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.        │
│  Agent:                                                                                                         │
│  Inventory Manager                                                                                              │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭───────────────────────────────────────────────── Crew Failure ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Failed                                                                                          │
│  Name:                                                                                                          │
│  crew                                                                                                           │
│  ID:                                                                                                            │
│  526a472c-7322-4412-8afa-4dd5ef5592ef                                                                           │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── Tracing Status ─────────────────────────────────────────────────╮
│                                                                                                                 │
│  Info: Tracing is disabled.                                                                                     │
│                                                                                                                 │
│  To enable tracing, do any one of these:                                                                        │
│  • Set tracing=True in your Crew/Flow code                                                                      │
│  • Set CREWAI_TRACING_ENABLED=true in your project's .env file                                                  │
│  • Run: crewai traces enable                                                                                    │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── Execution Traces ────────────────────────────────────────────────╮
│                                                                                                                 │
│  🔍 Detailed execution traces are available!                                                                    │
│                                                                                                                 │
│  View insights including:                                                                                       │
│    • Agent decision-making process                                                                              │
│    • Task execution flow and timing                                                                             │
│    • Tool usage details                                                                                         │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

Would you like to view your execution traces? [y/N] (20s timeout): 

AuthenticationError: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}

**Reasoning**:
The previous execution failed because of an incorrect API key provided in a preceding cell. I will re-provide the code to instantiate and execute the Crew, ensuring that it correctly follows the sequential process to manage the grocery shop workflow as intended, although the user will need to provide a valid OpenAI API key in the environment to avoid the 401 AuthenticationError.



In [6]:
from crewai import Crew, Process

# Instantiate the Crew with agents and tasks
grocery_crew = Crew(
    agents=[inventory_manager, personal_shopper],
    tasks=[check_inventory_task, personal_shopper_task],
    process=Process.sequential,
    verbose=True
)

# Execute the workflow
try:
    result = grocery_crew.kickoff()
    print("\n\n########################")
    print("## Final Result:")
    print("########################\n")
    print(result)
except Exception as e:
    print(f"An error occurred during execution: {e}")
    print("Please ensure a valid OPENAI_API_KEY is set in the environment variables.")

╭─────────────────────────────────────────── 🚀 Crew Execution Started ───────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Started                                                                                         │
│  Name:                                                                                                          │
│  crew                                                                                                           │
│  ID:                                                                                                            │
│  16b93577-58cf-4759-8b95-8688be316c9c                                                                           │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── 📋 Task Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Started                                                                                                   │
│  Name: Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.  │
│  ID: 138161f3-a087-4211-b77a-6d684bc4a480                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Inventory Manager                                                                                       │
│                                                                                                                 │
│  Task: Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.  │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

ERROR:root:OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
ERROR:root:OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}


╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided:            │
│  your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':        │
│  'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}                                            │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[CrewAIEventsBus] Warning: Event pairing mismatch. 'llm_call_failed' closed 'agent_execution_started' (expected 
'llm_call_started')

An unknown error occurred. Please check the details below.
Error details: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
An unknown error occurred. Please check the details below.
Error details: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}


[CrewAIEventsBus] Warning: Event pairing mismatch. 'agent_execution_error' closed 'task_started' (expected 
'agent_execution_started')

╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: OpenAI API call failed: Error code: 401 - {'error': {'message': 'Incorrect API key provided:            │
│  your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':        │
│  'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}                                            │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[CrewAIEventsBus] Warning: Event pairing mismatch. 'task_failed' closed 'crew_kickoff_started' (expected 
'task_started')

[CrewAIEventsBus] Warning: Ending event 'crew_kickoff_failed' emitted with empty scope stack. Missing starting 
event?

╭──────────────────────────────────────────────── 📋 Task Failure ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Failed                                                                                                    │
│  Name:                                                                                                          │
│  Check the availability of specific items like 'milk', 'eggs', and 'bread' in the store's current stock.        │
│  Agent:                                                                                                         │
│  Inventory Manager                                                                                              │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

An error occurred during execution: Error code: 401 - {'error': {'message': 'Incorrect API key provided: your_api*****here. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
Please ensure a valid OPENAI_API_KEY is set in the environment variables.


╭───────────────────────────────────────────────── Crew Failure ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Failed                                                                                          │
│  Name:                                                                                                          │
│  crew                                                                                                           │
│  ID:                                                                                                            │
│  16b93577-58cf-4759-8b95-8688be316c9c                                                                           │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯